## Part 1: List Comprehension Questions
### Q1: Given this list of temperatures in Celsius:
celsius_temps = [0, 20, 37, 100, -10, 15]
- Write a list comprehension to convert them to Fahrenheit using the formula: F = (C × 9/5) + 32


In [3]:
celsius_temps = [0, 20, 37, 100, -10, 15]
fahrenheit_temps = [(c * 9/5) + 32 for c in celsius_temps]
fahrenheit_temps

[32.0, 68.0, 98.6, 212.0, 14.0, 59.0]

### Q2: You have a list of customer ages:
ages = [25, 42, 17, 30, 65, 19, 21, 55, 16]
- Create a list comprehension that keeps only ages 18 and older.

In [10]:
ages = [25, 42, 17, 30, 65, 19, 21, 55, 16]
above_18 = [age for age in ages if age >=18] 
above_18

[25, 42, 30, 65, 19, 21, 55]

### Q3: You're analyzing sales data:
sales = [1250, 890, 2100, 540, 3100, 980]
- Create a list comprehension that returns "High" for sales > 1500, "Medium" for sales between 500-1500, and "Low" for sales < 500.

In [ ]:
sales = [1250, 890, 2100, 540, 3100, 980]
sales_categories = [('High' if s >1500 else 'Medium' if s >=500 else 'Low') for s in sales]

### Q4: Given this list of strings:
products = ["laptop", "mouse", "keyboard", "monitor", "printer"]
- Create a list comprehension that returns each product name capitalized and with " - In Stock" appended.



In [19]:
products = ["laptop", "mouse", "keyboard", "monitor", "printer"]
products_instock = [p.capitalize() + " - In Stock" for p in products]
products_instock

['Laptop - In Stock',
 'Mouse - In Stock',
 'Keyboard - In Stock',
 'Monitor - In Stock',
 'Printer - In Stock']

### Q5: You have a list with mixed data types:
data = [10, "20", 30.5, "invalid", 40, "50.5", "error"]
- Write a list comprehension that converts only the numeric strings to floats, leaving non-numeric strings as they are.

In [20]:
data = [10, "20", 30.5, "invalid", 40, "50.5", "error"]
data_2 = [
    float(item) if isinstance(item, str) and item.replace(".", "", 1).isdigit() 
    else item
    for item in data
]
data_2

[10, 20.0, 30.5, 'invalid', 40, 50.5, 'error']

## Part 2: Dictionary Comprehension Questions

## Q6: You have two lists:
- products = ["laptop", "mouse", "keyboard", "monitor"]
- prices = [999.99, 29.99, 79.99, 299.99]
- Create a dictionary comprehension that pairs each product with its price.

In [23]:
products = ["laptop", "mouse", "keyboard", "monitor"]
prices = [999.99, 29.99, 79.99, 299.99]
product_prices = {p: price for p, price in zip(products, prices)}
product_prices

{'laptop': 999.99, 'mouse': 29.99, 'keyboard': 79.99, 'monitor': 299.99}

## Q7: Using the dictionary from Q6, create a new dictionary with 15% discount applied only to items priced above $100.

In [25]:
discounts = {
    product: price * 0.85  
    for product, price in product_prices.items()
    if price > 100
}
discounts

{'laptop': 849.9915, 'monitor': 254.9915}

## Q8: From the sales_data list in your example:
Create a dictionary comprehension that shows for each product:

- Key: product name
- Value: tuple containing (price, units, revenue)

In [ ]:
sales_data = {
    "laptop": {"price": 999.99, "units": 5},
    "mouse": {"price": 29.99, "units": 30},
    "keyboard": {"price": 79.99, "units": 12},
    "monitor": {"price": 299.99, "units": 8}
}

product_summary = {
    product: (info["price"], info["units"], info["price"] * info["units"])
    for product, info in sales_data.items()}
product_summary

{'laptop': (999.99, 5, 4999.95),
 'mouse': (29.99, 30, 899.6999999999999),
 'keyboard': (79.99, 12, 959.8799999999999),
 'monitor': (299.99, 8, 2399.92)}

## Q9: You have a dictionary of student names and their scores:
scores = {"Alice": 88, "Bob": 72, "Charlie": 95, "Diana": 81}
Create a new dictionary that shows letter grades:

- A for scores ≥ 90
- B for scores 80-89
- C for scores 70-79
- D for scores < 70

In [34]:
scores = {"Alice": 88, "Bob": 72, "Charlie": 95, "Diana": 81}
grades = {
    name: (
        "A" if score >= 90 else
        "B" if score >= 80 else
        "C" if score >= 70 else
        "D"
    )
    for name, score in scores.items()
}
grades

{'Alice': 'B', 'Bob': 'C', 'Charlie': 'A', 'Diana': 'B'}

## Q10: Given this list of employee data dictionaries:

employees = [
    {"name": "John", "hours": 45, "rate": 25.50},
    {"name": "Sarah", "hours": 40, "rate": 32.00},
    {"name": "Mike", "hours": 50, "rate": 28.75}
]
- Create a dictionary comprehension where:

- Key: employee name
- Value: total pay (hours × rate) with overtime pay (1.5× rate) for hours over 40

In [36]:
employees = [
    {"name": "John", "hours": 45, "rate": 25.50},
    {"name": "Sarah", "hours": 40, "rate": 32.00},
    {"name": "Mike", "hours": 50, "rate": 28.75}
]

total_pay = {
    emp["name"]: (
        emp["hours"] * emp["rate"] if emp["hours"] <= 40
        else (40 * emp["rate"]) + ((emp["hours"] - 40) * emp["rate"] * 1.5)
    )
    for emp in employees}
total_pay

{'John': 1211.25, 'Sarah': 1280.0, 'Mike': 1581.25}

## Part 3: Combined Comprehension Questions
### Q11: Using the sales_data list from your example, create a list comprehension that returns only the products with revenue greater than $300.


In [38]:
high_revenue = [
    product
    for product, info in sales_data.items()
    if info["price"] * info["units"] > 300
]
high_revenue

['laptop', 'mouse', 'keyboard', 'monitor']

### Q12: Create a dictionary from the sales_data where:

- Key: product name
- Value: percentage of total revenue that product contributes

In [41]:
total_rev = sum(info["price"] * info["units"] for info in sales_data.values())
revenue_percentage = {
    product: (info["price"] * info["units"]) / total_rev * 100
    for product, info in sales_data.items()
}
revenue_percentage

{'laptop': 53.998347634038744,
 'mouse': 9.716559838867319,
 'keyboard': 10.366490450296723,
 'monitor': 25.91860207679722}

### Q13: Given this data:

customers = [
    {"name": "Alice", "purchases": [50, 125, 75]},
    {"name": "Bob", "purchases": [200, 50]},
    {"name": "Charlie", "purchases": [300, 150, 100, 50]}
]
- Create a dictionary comprehension where:

- Key: customer name
- Value: total amount spent

In [43]:
customers = [
    {"name": "Alice", "purchases": [50, 125, 75]},
    {"name": "Bob", "purchases": [200, 50]},
    {"name": "Charlie", "purchases": [300, 150, 100, 50]}
]

customer_totals = {
    cust["name"]: sum(cust["purchases"])
    for cust in customers
}
customer_totals

{'Alice': 250, 'Bob': 250, 'Charlie': 600}

### Q14: From the customers list in Q13, create a list comprehension that returns customer names who have spent more than $400 in total.

In [44]:
big_spenders = [
    cust["name"]
    for cust in customers
    if sum(cust["purchases"]) > 400
]
big_spenders

['Charlie']

### Q15: You're analyzing website traffic data:

daily_visitors = {
    "Monday": [1200, 1100, 1300, 1400],
    "Tuesday": [1000, 900, 1100, 1200],
    "Wednesday": [1500, 1400, 1600, 1700]
}
- Create a dictionary comprehension that shows for each day:

- Key: day name
- Value: average visitors (sum/4)

In [45]:
daily_visitors = {
    "Monday": [1200, 1100, 1300, 1400],
    "Tuesday": [1000, 900, 1100, 1200],
    "Wednesday": [1500, 1400, 1600, 1700]
}

daily_avg = {
    day: sum(visitors) / len(visitors)
    for day, visitors in daily_visitors.items()
}
daily_avg



{'Monday': 1250.0, 'Tuesday': 1050.0, 'Wednesday': 1550.0}

## Part 4: Challenge Questions
### Q16: The company is running a promotion: "Buy 2, get 1 free" on all items.
- Using your original sales_data, calculate the new revenue if this promotion was applied to all sales.


In [46]:
promo_revenue = {
    product: (
        ((info["units"] // 3) * 2 + (info["units"] % 3)) * info["price"]
    )
    for product, info in sales_data.items()
}
promo_revenue

{'laptop': 3999.96, 'mouse': 599.8, 'keyboard': 639.92, 'monitor': 1799.94}

### Q17: You need to create an inventory report that shows:

- Products sorted by profitability (revenue per unit sold)
Mark each product as "Restock Needed" if units sold > 10

In [ ]:
inventory_report = [
    {
        "product": product,
        "profitability": info["price"],
        "status": "Restock Needed" if info["units"] > 10 else "OK"
    }
    for product, info in sorted(
        sales_data.items(),
        key=lambda x: x[1]["price"],         
        reverse=True
    )
]
inventory_report

[{'product': 'laptop', 'profitability': 999.99, 'status': 'OK'},
 {'product': 'monitor', 'profitability': 299.99, 'status': 'OK'},
 {'product': 'keyboard', 'profitability': 79.99, 'status': 'Restock Needed'},
 {'product': 'mouse', 'profitability': 29.99, 'status': 'Restock Needed'}]

### Q18: Create a comprehensive sales summary dictionary using only comprehensions that includes:

- Total revenue for each product
- Average price across all products
- Total units sold
- Most profitable product
- List of products that need restocking (units > 10)

In [48]:
# Total revenue per product
product_revenue = {
    p: info["price"] * info["units"]
    for p, info in sales_data.items()
}

# Average price
avg_price = sum(info["price"] for info in sales_data.values()) / len(sales_data)

# Total Units sold
total_units = sum(info["units"] for info in sales_data.values())

# Most profitable product
most_profitable = max(product_revenue, key=product_revenue.get)

# Products needing restock(units>10)
restock_items_list = [p for p, info in sales_data.items() if info["units"] > 10]


### Q19: Write a single dictionary comprehension that transforms sales_data into:

{
    "Laptop": {
        "price": 999.99,
        "units": 5,
        "revenue": 4999.95,
        "category": "Electronics"
    },
    # ... and so on for other products
}
- (Assume categories: Laptop="Electronics" Mouse="Accessories", Keyboard="Accessories", Monitor="Electronics")

In [49]:
categories = {
    "laptop": "Electronics",
    "mouse": "Accessories",
    "keyboard": "Accessories",
    "monitor": "Electronics"
}

detailed_sales = {
    p.capitalize(): {
        "price": info["price"],
        "units": info["units"],
        "revenue": info["price"] * info["units"],
        "category": categories[p]
    }
    for p, info in sales_data.items()
}
detailed_sales

{'Laptop': {'price': 999.99,
  'units': 5,
  'revenue': 4999.95,
  'category': 'Electronics'},
 'Mouse': {'price': 29.99,
  'units': 30,
  'revenue': 899.6999999999999,
  'category': 'Accessories'},
 'Keyboard': {'price': 79.99,
  'units': 12,
  'revenue': 959.8799999999999,
  'category': 'Accessories'},
 'Monitor': {'price': 299.99,
  'units': 8,
  'revenue': 2399.92,
  'category': 'Electronics'}}

### Q20: Create a function that takes any sales data list and returns a complete analysis using comprehensions only. The function should return a dictionary with all the metrics you've calculated.

In [50]:
def analyze_sales(data):
    product_revenue = {
        p: info["price"] * info["units"]
        for p, info in data.items()
    }

    avg_price = (
        sum(info["price"] for info in data.values()) / len(data)
        if len(data) > 0 else 0
    )

    total_units = sum(info["units"] for info in data.values())

    most_profitable = (
        max(product_revenue, key=product_revenue.get)
        if len(product_revenue) > 0 else None
    )

    restock_needed = [
        p for p, info in data.items() if info["units"] > 10
    ]

    return {
        "product_revenue": product_revenue,
        "average_price": avg_price,
        "total_units_sold": total_units,
        "most_profitable_product": most_profitable,
        "restock_needed": restock_needed
    }